machine_management/creating-infrastructure-machinesets.adoc

```diff
@@ -62,6 +62,22 @@ include::modules/machineset-yaml-vsphere.adoc[leveloffset=+3]
 
 include::modules/machineset-creating.adoc[leveloffset=+2]
 
+[id="assigning-machineset-resources-to-infra-nodes"]
+== Assigning MachineSet resources to infrastructure nodes
+
+After creating an infrastructure MachineSet, the `worker` and `infra` roles are applied to new infra nodes. The Machine Config Operator (MCO) requires the `worker` role for infra nodes so it can manage the lifecycle of the node. Nodes with the `infra` role applied are not counted toward the total number of subscriptions that are required to run the environment, even when the `worker` role is also applied.
```

Comment:
`The Machine Config Operator (MCO) requires the worker role for infra nodes so it can manage the lifecycle of the node.`

I don't believe that's true today. You can remove the worker label from the node and the MCO will still update it as if it was a worker, since the infra pool inherits from the worker pool.

Reply:
@yuqi-zhang do you have any idea why this is still being done if the MCO doesn't require it?

Reply:
There are two reasons I think:
1. you cannot bring up a node as an "infra node" since you need to provide a secret for the node's ignition to fetch its config. By default our clusters only have "master" and "worker" secrets which means the node must boot as a worker or a master. Although now I think about it I think there's no reason why the machineset definition can't specify it as a infra node but bootstrap with a worker secret. It should immediately perform an update to the correct infra config.

2. It used to be that you needed both labels otherwise the node would degrade, but that's been fixed since sometime earlier this year. We've always created nodes as such and I suppose nobody took the time to investigate whether we could bootstrap infra nodes independently.

Reply:
Thanks for those details! I will update this text so it's accurate.

Reply:
@yuqi-zhang looking back to this as there have been some complaints and questions for why the `worker` label can't just be removed after it's initialized. Do you know what version of OCP first began allowing the `worker` label to be removed without degradation? And any BZs/JIRAs you can't refer to for that fix? It may be worth updating this content with a different procedure for cluster versions that allow it, since it sounds like taints/tolerations have more restrictions.

Reply:
we updated the docs in our repo in june showing the _option_ to remove the worker label: https://github.com/openshift/machine-config-operator/pull/1824

@runcom isn't around this week but he'd be the better person to ask

Reply:
@kikisdeliveryservice thanks for the info! The big dilemma here is what path to promote for workload assignments on an infra node:
- taints/tolerations leaving both `infra`/`worker` roles applied to an infra node
- removing the `worker` role, as suggested in [custom-pools.md](https://github.com/openshift/machine-config-operator/blob/master/docs/custom-pools.md)

It seems that just removing the `worker` label is easiest, but that caused problems in earlier versions of OCP (e.g., upgrading issues). What I'm trying to figure out is if the `worker` label is now the bonafide, absolute recommendation (with the assumption that issues have been resolved), or if there are still deficiencies for doing this, and risks we would need to document for customers going that route. Or perhaps we just document both and let customers pick what they want?

Reply:
AFAIK, unless something has changed, the instructions in our repo are the supported path as @yuqi-zhang stated. I'd consider anything that caused an upgrade to fail in recent ocp versions (since the docs were updated in june) to be a bug. I'd have to talk to @runcom to get his thoughts on which version should have which instructions but he's out rn... happy to get clarity on this.

Reply:
@kikisdeliveryservice thank you! Let me know what comes of this discussion. It has been hard to get this nailed down, but hopefully @runcom can give his final say and get this cleared up.